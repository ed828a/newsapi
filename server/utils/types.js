const Categories = Object.freeze({
  Business: "business",
  Entertainment: "entertainment",
  General: "general",
  Health: "health",
  Science: "science",
  Sports: "sports",
  Technology: "technology",
  Topheadline: "topheadline"
});

export { Categories };
