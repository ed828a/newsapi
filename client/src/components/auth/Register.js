import React, { Component, Fragment } from "react";
import axios from "axios";
import config from "../../config";
import { Link } from "react-router-dom";

const url = `${config.BASE_URL}/api/auth/register`;

export default class Register extends Component {
  state = {
    username: "",
    password: "",
    error: "",
    validLogin: false
  };

  handleChange = event => {
    this.setState({
      [event.target.name]: event.target.value
    });
  };
  handleSubmit = event => {
    axios
      .post(url, {
        username: this.state.username,
        password: this.state.password
      })
      .then(data => {
        console.log(data);
        this.setState({
          validLogin: true,
          username: "",
          password: "",
          error: ""
        });
      })
      .catch(err => {
        this.setState(error => {
          const errorkey = Object.keys(err.response.data.errors);
          const errorval = Object.values(err.response.data.errors);
          const msg = errorkey[0] + " " + errorval[0];
          return { error: msg };
        });
      });

    event.preventDefault();
  };

  render() {
    return (
      <div id="main">
        <div className="container">
          <div className="row">
            <div className="marginleft33">
              <h1>Register</h1>
            </div>
          </div>

          {this.state.error ? (
            <Fragment>
              <div className="row">&nbsp;</div>
              <div className="row marginleft40" style={{ color: "red" }}>
                {this.state.error}
              </div>
            </Fragment>
          ) : (
            ""
          )}
          {this.state.validLogin ? (
            <Fragment>
              <div className="row">&nbsp;</div>

              <div className="marginleft40" style={{ color: "green" }}>
                User successFully Registered.Please log in &nbsp;
                <Link to="/login" style={{ textDecoration: "underline" }}>
                  <b>here</b>
                </Link>
                <div className="row">&nbsp;</div>
              </div>
            </Fragment>
          ) : (
            ""
          )}
          <div className="row">
            <div className="marginleft33">Username :</div>
            <div className="cols-md-4">
              <input
                type="text"
                name="username"
                value={this.state.username}
                onChange={this.handleChange}
              />
            </div>
          </div>
          <div className="row">
            <div className="marginleft33">Password : </div>
            <div className="cols-md-4">
              <input
                type="password"
                name="password"
                value={this.state.password}
                onChange={this.handleChange}
              />
            </div>
          </div>

          <div className="row">
            <div className="marginleft40" />
            <div className="cols-md-4">
              <input
                className="btnlogin"
                type="button"
                name="button"
                value="Register"
                onClick={this.handleSubmit}
              />
            </div>
          </div>
        </div>
      </div>
    );
  }
}
