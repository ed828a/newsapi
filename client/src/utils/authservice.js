import axios from "axios";
import decode from "jwt-decode";
import config from "../config";
class AuthService {
  url = `${config.BASE_URL}/api/auth/login`;
  login = (username, password) => {
    return axios
      .post(this.url, {
        username,
        password
      })
      .then(res => res.data)
      .then(res => {
        this.setToken(res.user.token);
        return Promise.resolve(res);
      })
      .catch(err => Promise.reject(err));
  };
  setToken(token) {
    axios.defaults.headers.common["Authorization"] = token;
    localStorage.setItem("jwttoken", token);
  }
  isTokenExpired(token) {
    try {
      const decoded = decode(token);
      console.log(decoded);
      console.log(Date.now());
      return decoded.exp < Date.now() / 1000 ? true : false;
    } catch (err) {
      return false;
    }
  }
  isAuthenticated() {
    const token = this.getToken();
    console.log(token);
    if (!token || this.isTokenExpired(token)) {
      return false;
    } else {
      return true;
    }
  }
  getToken() {
    return localStorage.getItem("jwttoken");
  }

  logout() {
    delete axios.defaults.headers.common["Authorization"];
    // Clear user token and profile data from localStorage
    localStorage.removeItem("jwttoken");
  }
}

export let authService = new AuthService();
