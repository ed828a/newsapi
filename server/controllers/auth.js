import User from "../models/user";
import passport from "passport";
import {
  userToJSON,
  getUserToken,
  encryptPassword,
  isPasswordAllowed
} from "../utils/auth";

const authUserJson = user => ({
  ...userToJSON(user),
  token: getUserToken(user)
});
async function register(req, res) {
  const { username, password } = req.body;
  if (!username) {
    return res.status(422).json({ errors: { username: `can't be blank` } });
  }
  if (!password) {
    return res.status(422).json({ errors: { password: `can't be blank` } });
  }

  if (!isPasswordAllowed(password)) {
    return res.status(422).json({
      errors: {
        password: ` must be greater than 6 character.Must contain special
      character and letter`
      }
    });
  }

  const existingUser = await User.findOne({ username });

  if (existingUser) {
    return res.status(422).json({ errors: { username: "taken" } });
  }
  const encPassword = await encryptPassword(password);

  const newUser = await User.create({
    username,
    password: encPassword
  });

  return res.json({ user: authUserJson(newUser) });
}

async function login(req, res, next) {
  const { username, password } = req.body;
  if (!username) {
    return res.status(422).json({ errors: { username: `can't be blank` } });
  }
  if (!password) {
    return res.status(422).json({ errors: { password: `can't be blank` } });
  }

  const { user, info } = await authenticate(req, res, next);
  if (user) {
    return res.json({ user: authUserJson(user) });
  } else {
    return res.status(422).json(info);
  }
}
function authenticate(req, res, next) {
  return new Promise((resolve, reject) => {
    passport.authenticate("local", { session: false }, (err, user, info) => {
      if (err) {
        reject(err);
      } else {
        resolve({ user, info });
      }
    })(req, res, next);
  });
}
async function me(req, res) {}

export { me, login, register };
