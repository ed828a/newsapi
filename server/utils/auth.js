import jwt from "jsonwebtoken";
import bcrypt from "bcryptjs";
import config from "../config";
import expressJWT from "express-jwt";
import LocalStrategy from "passport-local";
import User from "../models/user";
import { omit } from "lodash";
const twoHours = 60 * 60 * 24;
const requireTime = Date.now();

const now = () =>
  process.env.NODE_ENV === "production" ? Date.now() : requireTime;
const issuedAt = Math.floor(now() / 1000);

const authMiddleware = {
  required: expressJWT({
    secret: config.secret
  }),
  optional: expressJWT({
    secret: config.secret,
    credentialsRequired: false
  })
};

async function encryptPassword(plainText) {
  const salt = await bcrypt.genSalt(10);
  const encPassword = await bcrypt.hash(plainText, salt);
  return encPassword;
}

async function comparePassword(plainPassword, encryptedPassword) {
  try {
    const result = await bcrypt.compare(plainPassword, encryptedPassword);
    return result;
  } catch (error) {
    return false;
  }
}
function isPasswordAllowed(password) {
  return password.length > 6 && /\d/.test(password) && /\D/.test(password);
}

function getLocalStrategy() {
  return new LocalStrategy(async (username, password, done) => {
    let user;

    try {
      user = await User.findOne({ username }).lean();
    } catch (error) {
      return done(error);
    }

    if (!user || (await comparePassword(password, user.password)) === false) {
      return done(null, false, {
        errors: { "username or password": "is invalid" }
      });
    }

    return done(null, userToJSON(user));
  });
}

function getUserToken({ _id, username }) {
  const issuedAt = Math.floor(now());
  return jwt.sign(
    {
      id: _id,
      username,
      iat: issuedAt,
      exp: issuedAt + twoHours
    },
    config.secret
  );
}
function userToJSON(user) {
  return omit(user, ["createdAt", "password", "__v", "isActive"]);
}
export {
  authMiddleware,
  getUserToken,
  getLocalStrategy,
  isPasswordAllowed,
  encryptPassword,
  comparePassword,
  userToJSON
};
