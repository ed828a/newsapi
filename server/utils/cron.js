import cron from "node-cron";
import Article from "../models/article";
import { Categories } from "./types";
import NewsAPI from "newsapi";
import config from "../config";

const newsapi = new NewsAPI(config.API_KEY);

function insertArticles(articles, category = Categories.Topheadline) {
  let promises = articles.map(article => {
    let doc = {
      category,
      source: article.source.name,
      author: article.author,
      description: article.description,
      url: article.url,
      urltoimage: article.urlToImage,
      publishedAt: article.publishedAt,
      content: article.content
    };
    return Article.create(doc)
      .then(data => {
        console.log("successfully updated");
      })
      .catch(err => {
        console.log(err.message);
      });
  });
  return promises;
}
function categoryNewsCron(category) {
  //return new Promise((resolve, reject) => {
  cron.schedule("*/20 * * * *", () => {
    console.log("running every minute 15 for category" + category);
    Article.remove({ category })
      .then(() => {
        newsapi.v2
          .topHeadlines({
            category,
            language: "en",
            country: "us"
          })
          .then(response => {
            if (response.articles && response.articles.length > 0) {
              let promises = insertArticles(response.articles, category);
              Promise.all(promises)
                .then(() => console.log("cron completed successfully "))
                .catch(err => {
                  console.log("failed", err);
                });
              //resolve(promises);
            }
          });
      })
      .catch(err => {
        console.log(err);
        // reject(err);
      });
  });
  // });
}

function topNewsDataCron() {
  cron.schedule("*/10 * * * *", () => {
    console.log("running a task every 5 minute");
    Article.remove({ category: Categories.Topheadline })
      .then(() => {
        newsapi.v2
          .topHeadlines({
            language: "en",
            country: "us"
          })
          .then(response => {
            console.log(response.articles);
            if (response.articles && response.articles.length > 0) {
              let promises = insertArticles(response.articles);
              Promise.all(promises).then(() =>
                console.log("cron completed successfully")
              );
            }
          });
      })
      .catch(err => console.log(err));
  });
}

function setUpCronJob() {
  topNewsDataCron();
  const categories = Object.values(Categories);
  categories.map(category => {
    categoryNewsCron(category);
  });
}

export default setUpCronJob;
