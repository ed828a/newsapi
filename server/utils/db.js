import mongoose from "mongoose";
import settings from "../config";
mongoose.Promise = global.Promise;

export const connect = () =>
  mongoose
    .connect(
      settings.mongoURI,
      { useNewUrlParser: true }
    )
    .then(() => {
      console.log("connected to mongo");
    })
    .catch(err => {
      console.log("connection error" + err);
    });
