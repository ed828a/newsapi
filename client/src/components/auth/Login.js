import React, { Component, Fragment } from "react";
import { authService } from "../../utils/authservice";

export default class Login extends Component {
  state = {
    username: "",
    password: "",
    error: ""
  };

  handleChange = event => {
    this.setState({
      [event.target.name]: event.target.value
    });
  };
  handleSubmit = event => {
    authService
      .login(this.state.username, this.state.password)
      .then(data => this.props.history.push("/"))
      .catch(err => {
        this.setState(error => {
          const errorkey = Object.keys(err.response.data.errors);
          const errorval = Object.values(err.response.data.errors);
          const msg = errorkey[0] + " " + errorval[0];
          return { error: msg };
        });
      });

    event.preventDefault();
  };

  render() {
    return (
      <div id="main">
        <div className="container">
          <div className="row">
            <div className="marginleft33">
              <h1>Login</h1>
            </div>
          </div>

          {this.state.error ? (
            <Fragment>
              <div className="row">&nbsp;</div>
              <div className="row marginleft40" style={{ color: "red" }}>
                {this.state.error}
              </div>
            </Fragment>
          ) : (
            ""
          )}
          <div className="row">
            <div className="marginleft33">Username :</div>
            <div className="cols-md-4">
              <input
                type="text"
                name="username"
                value={this.state.username}
                onChange={this.handleChange}
              />
            </div>
          </div>
          <div className="row">
            <div className="marginleft33">Password : </div>
            <div className="cols-md-4">
              <input
                type="password"
                name="password"
                value={this.state.password}
                onChange={this.handleChange}
              />
            </div>
          </div>
          <div className="row">
            <div className="marginleft40" />
            <div className="cols-md-4">
              <input
                className="btnlogin"
                type="button"
                name="button"
                value="Login"
                onClick={this.handleSubmit}
              />
            </div>
          </div>
        </div>
      </div>
    );
  }
}
