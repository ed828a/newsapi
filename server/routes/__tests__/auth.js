import axios from "axios";
import faker from "faker";
import { getUserToken, encryptPassword } from "../../utils/auth";

import startServer from "../../start";
import User from "../../models/user";

describe("authentication : ", () => {
  let server, baseURL, api;
  beforeAll(async () => {
    server = await startServer({ port: 8778 });
    baseURL = `http://localhost:${server.address().port}/api`;
    api = axios.create({ baseURL });
    await User.remove();
  });

  afterAll(async () => await server.close());

  describe("Login", () => {
    test("should require  username to login", async () => {
      const error = await api
        .post("/auth/login", {
          password: faker.internet.password()
        })
        .catch(err => err.response);
      expect(error).toMatchObject({
        status: 422,
        data: { errors: { username: expect.any(String) } }
      });
      expect(error.data.errors).toMatchSnapshot();
    });

    test("should require password to login", async () => {
      const error = await api
        .post("auth/login", { username: faker.internet.userName() })
        .catch(err => err.response);
      expect(error).toMatchObject({
        status: 422,
        data: { errors: { password: expect.any(String) } }
      });
      expect(error.data.errors).toMatchSnapshot();
    });
    test("should not login with a non user", async () => {
      const error = await api
        .post("auth/login", {
          username: faker.internet.userName(),
          password: faker.internet.password()
        })
        .catch(err => err.response);
      expect(error).toMatchObject({
        status: 422,
        data: { errors: { "username or password": expect.any(String) } }
      });
      expect(error.data.errors).toMatchSnapshot();
    });
    test("should login with valid user", async () => {
      const password = faker.internet.password();
      const encPassword = await encryptPassword(password);
      const userObj = {
        username: faker.internet.userName(),
        password: encPassword
      };
      const newUser = await User.create(userObj);

      const user = await api
        .post("auth/login", {
          username: newUser.username,
          password: password
        })
        .then(res => {
          return res.data.user;
        })
        .catch(err => console.log(err.response));

      expect(user).toEqual({
        token: expect.any(String),
        _id: newUser.id,
        username: newUser.username
      });
      await User.deleteOne({ _id: newUser.id });
    });
  });
  describe("register", () => {
    test("should require username to register", async () => {
      const error = await api
        .post("/auth/register", {
          password: faker.internet.password()
        })
        .catch(err => err.response);
      expect(error).toMatchObject({
        status: 422,
        data: { errors: { username: expect.any(String) } }
      });
      expect(error.data.errors).toMatchSnapshot();
    });
    test("should require password to register", async () => {
      const error = await api
        .post("auth/register", { username: faker.internet.userName() })
        .catch(err => err.response);
      expect(error).toMatchObject({
        status: 422,
        data: { errors: { password: expect.any(String) } }
      });
      expect(error.data.errors).toMatchSnapshot();
    });
    test("should require unique username", async () => {
      const password = faker.internet.password();

      const userObj = {
        username: faker.internet.userName(),
        password
      };
      const newUser = await User.create(userObj);
      const error = await api
        .post("/auth/register", userObj)
        .catch(err => err.response);

      expect(error).toMatchObject({
        status: 422,
        data: { errors: { username: expect.any(String) } }
      });
      expect(error.data.errors).toMatchSnapshot();

      await User.deleteOne({ _id: newUser.id });
    });
    test("should allow register with correct username password", async () => {
      const userObj = {
        username: faker.internet.userName(),
        password: faker.internet.password()
      };

      const user = await api
        .post("/auth/register", userObj)
        .then(res => res.data.user);

      expect(user).toMatchObject({
        username: userObj.username,
        password: expect.any(String)
      });

      await User.deleteOne({ _id: user._id });
    });
  });
});
