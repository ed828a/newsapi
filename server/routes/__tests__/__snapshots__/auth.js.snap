// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authentication :  Login should not login with a non user 1`] = `
Object {
  "username or password": "is invalid",
}
`;

exports[`authentication :  Login should require  username to login 1`] = `
Object {
  "username": "can't be blank",
}
`;

exports[`authentication :  Login should require password to login 1`] = `
Object {
  "password": "can't be blank",
}
`;

exports[`authentication :  register should require password to register 1`] = `
Object {
  "password": "can't be blank",
}
`;

exports[`authentication :  register should require unique username 1`] = `
Object {
  "username": "taken",
}
`;

exports[`authentication :  register should require username to register 1`] = `
Object {
  "username": "can't be blank",
}
`;
